File: 3822-fill-a-special-grid/3822-fill-a-special-grid.py
class Solution:
    def specialGrid(self, N: int) -> List[List[int]]:
        size = 1 << N
        grid = [[0] * size for _ in range(size)]
        val = 0

        def solve(rowStart, rowEnd, colStart, colEnd):
            nonlocal val
            sz = rowEnd - rowStart
            if sz == 1:
                grid[rowStart][colStart] = val
                val += 1
                return
            midRow = rowStart + sz // 2
            midCol = colStart + sz // 2

            solve(rowStart, midRow, midCol, colEnd) # top right
            solve(midRow, rowEnd, midCol, colEnd) # bottom right
            solve(midRow, rowEnd, colStart, midCol) # bottom left
            solve(rowStart, midRow, colStart, midCol) # top left

        solve(0, size, 0, size)
        return grid

File: 3822-fill-a-special-grid/3822-fill-a-special-grid.java
class Solution {
    public int[][] specialGrid(int n) {
        int size = 1 << n;
        int[][] grid = new int[size][size];
        solve(grid, 0, 0, size);
        return grid;
    }
    private int val = 0;
    private void solve(int[][] grid, int r, int c, int size){
        if (size == 1){
            grid[r][c] = val++;
            return;
        }
        int s = size / 2;
        solve(grid, r, c + s, s); // top right
        solve(grid, r + s, c + s, s); // bottom right
        solve(grid, r + s, c, s); // bottom left
        solve(grid, r, c, s); // top left
    }
}